File: LeetCode_Algorithm/268/268.js
var missingNumber = function(nums) {
  
  // sol1 O(n), O(1) 65~98ms 92~54%
  // gauss summation
  // var length = nums.length
  // sumNums = parseInt(length*(length+1) / 2)
  // var currSum = nums.reduceRight(function(sum,currVal){
  //   return (sum+currVal)}, 0)
  // return sumNums-currSum

  // sol2 O(n), O(n) 80~138ms 77~21%
  // set과 hashmap
  // hashMap = new Set(nums)
  // for (let idx=0; idx<=nums.length; idx++){
  //   if (!hashMap.has(idx)){return idx}
  // }

  // sol3 O(n), O(1) 72~132ms 86~24%
  // 비트 조작
  var missing = nums.length
  // ES6
  // for (const [idx, num] of nums.entries()) {
  //   missing ^= idx ^ num
  // }
  // return missing
  
  // ES5 76~123ms 81~29%
  const arr = nums.map(function(num, idx){
    return [idx, num]
  })
  for (const [idx, num] of arr){
    missing ^= idx ^ num
  }
  return missing


  
};
console.log(missingNumber([2,0,1]))

File: LeetCode_Algorithm/268/268.py
import enum
from typing import List

class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        # sol1 O(n), O(1) 158~203ms 73 46%
        # n의 길이 라면 n(n+1)/2 
        # 배열을 돌면서 뺀 값을 리턴
        # n = len(nums)
        # sum_len_nums = int(n*(n+1)/2)
        # return sum_len_nums-sum(num for num in nums)

        # sol2 O(nlogn) O(1) 171~213ms 66~43%
        # sort
        # nums.sort()
        # for i in range(len(nums)):
        #     if i != nums[i]: return i
        # return len(nums)

        # sol3 O(n), O(n)
        # hashSet
        # hashSet = set(nums)
        # n = len(nums)+1
        # for number in range(n):
        #     if number not in hashSet: return number

        # sol4 O(n), O(1) 206 47
        # bit manipulation
        # 인덱스와 숫자간의 비트 조작을 통해 문제를 해결할 수도 있다.
        
        # n ^ n = 0
        # 0 ^ n = n
        # idx : 0 1 2 3
        # num : 0 1 3 4
        # total length : 4
        # (0^0) ^ (1^1) ^ (3^3) ^ (4^4) ^ 2 = 0 ^ 0 ^ 0 ^ 0 ^ 2
        
        # 전체 인덱스 + 전체 숫자 + 총 길이를 모두 xor 연산한다.
        ans = len(nums)
        for idx, num in enumerate(nums):
            ans ^= idx ^ num
        return ans







print(Solution().missingNumber([3,0,1]))